client/src/components/pages/Players.js

```diff
@@ -1,289 +1,303 @@
+        tourney: tourney,
+        regTime: getStatsById(_id).regTime,
+      });
+
+      setPlayers(
+        players.map((t) => {
+          if (t._id === _id) return newPlayer;
+
+          return t;
+        })
+      );
+    } catch {
+      message.error(`Something went wrong: ${e}`);
+    }
   };
 
-  handleSortChange = (e) => {
+  const handleSortChange = (e) => {
     const sort = e.target.value;
-    if (this.state.mode === "players") {
-      this.sortPlayers(sort);
+    if (mode === "players") {
+      setPlayers(sortedPlayers(sort, players));
     } else {
-      this.sortTeams(sort);
+      setTeams(sortedTeams(sort, teams));
     }
 
-    this.setState({ sort });
+    setSort(sort);
   };
 
-  handleAddPlayer = async () => {
-    this.setState({ modalLoading: true });
-    const player = await post("/api/force-register", {
-      ...this.state.addPlayerData,
-      tourney: this.props.tourney,
-    });
-    this.setState(
-      (state) => ({
-        players: [...state.players, player],
-        modalLoading: false,
-        modalVisible: false,
-      }),
-      () => this.sortPlayers(this.state.sort)
-    );
+  const handleAddPlayer = async () => {
+    setModalLoading(true);
+
+    try {
+      const playerData = await post("/api/force-register", {
+        ...addPlayerData,
+        tourney: tourney,
+      });
+
+      setPlayers(sortedPlayers(sort, [...players, playerData]));
+      setModalLoading(false);
+      setModalVisible(false);
+
+      message.success(`${playerData.username} added!`);
+    } catch {
+      setModalLoading(false);
+      message.error("Player not found");
+    }
   };
 
-  isAdmin = () => hasAccess(this.props.user, this.props.tourney, []);
+  const isAdmin = () => hasAccess(user, tourney, []);
 
-  refreshRanks = async () => {
-    this.setState({ refreshPercent: 0 });
+  const refreshRanks = async () => {
+    setRefreshPercent(0);
 
     let offset = 0;
-    while (offset < this.state.players.length) {
+    while (offset < players.length) {
       console.log(`Refreshing players at offset ${offset}`);
-      const result = await post("/api/refresh", { tourney: this.props.tourney, offset });
+      const result = await post("/api/refresh", { tourney: tourney, offset });
       offset = result.offset;
 
-      this.setState(
-        (state) => ({
-          players: state.players.map((p) => {
-            const newPlayer = result.players.filter((r) => r._id === p._id)[0];
-            return newPlayer || p;
-          }),
-          refreshPercent: Math.min(100, Math.round((100 * offset) / this.state.players.length)),
-        }),
-        () => this.sortPlayers(this.state.sort)
+      setPlayers(
+        players.map((p) => {
+          const newPlayer = result.players.filter((r) => r._id === p._id)[0];
+          return newPlayer || p;
+        })
       );
+      setRefreshPercent(Math.min(100, Math.round((100 * offset) / players.length)));
+      setPlayers(sortedPlayers(sort, players));
```

Comment:
doing these two `setPlayers` calls back to back will run into the same async issue as before where it might not use the updated players

combine the two `setPlayers` calls, for example:
```js
setPlayers(
  sortedPlayers(sort,
    players.map((p) => {
      const newPlayer = result.players.filter((r) => r._id === p._id)[0];
      return newPlayer || p;
    })
  )
);
```